odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2336 @@
+        of the weighting and the space must match.
+
+    See Also
+    --------
+    NumpyTensorSpaceConstWeighting
+    NumpyTensorSpaceArrayWeighting
+    """
+    return _weighting(weights, exponent=exponent).norm
+
+
+def npy_weighted_dist(weights, exponent=2.0, use_inner=False):
+    """Weighted distance on `TensorSpace`'s as free function.
+
+    Parameters
+    ----------
+    weights : scalar or `array-like`
+        Weights of the distance. A scalar is interpreted as a
+        constant weight, a 1-dim. array as a weighting vector.
+    exponent : positive `float`
+        Exponent of the norm.
+    use_inner : `bool`, optional
+        Calculate ``dist`` using the formula
+
+            ``||x - y||^2 = ||x||^2 + ||y||^2 - 2 * Re <x, y>``
+
+        This avoids the creation of new arrays and is thus faster
+        for large arrays. On the downside, it will not evaluate to
+        exactly zero for equal (but not identical) ``x`` and ``y``.
+
+        Can only be used if ``exponent`` is 2.0.
+
+    Returns
+    -------
+    dist : `callable`
+        Distance function with given weight. Constant weightings
+        are applicable to spaces of any size, for arrays the sizes
+        of the weighting and the space must match.
+
+    See Also
+    --------
+    NumpyTensorSpaceConstWeighting
+    NumpyTensorSpaceArrayWeighting
+    """
+    return _weighting(weights, exponent=exponent,
+                      dist_using_inner=use_inner).dist
+
+
+def _norm_default(x):
+    """Default Euclidean norm implementation."""
+    if _blas_is_applicable(x.data):
+        nrm2 = linalg.blas.get_blas_funcs('nrm2', dtype=x.dtype)
+        norm = partial(nrm2, n=native(x.size))
+    else:
+        norm = np.linalg.norm
+    return norm(x.data.ravel(order=x.order))
+
+
+def _pnorm_default(x, p):
+    """Default p-norm implementation."""
+    return np.linalg.norm(x.data.ravel(order=x.order), ord=p)
+
+
+def _pnorm_diagweight(x, p, w):
+    """Diagonally weighted p-norm implementation."""
+    # This is faster than first applying the weights and then summing with
+    # BLAS dot or nrm2
+    xp = np.abs(x.data.ravel(order=x.order))
+    if np.isfinite(p):
+        xp = np.power(xp, p, out=xp)
+        xp *= w.ravel(order='K')  # w is a plain NumPy array
+        return np.sum(xp) ** (1 / p)
+    else:
+        xp *= w.ravel(order='K')
+        return np.max(xp)
+
+
+def _inner_default(x1, x2):
+    """Default Euclidean inner product implementation."""
+    if _blas_is_applicable(x1.data, x2.data):
+        dotc = linalg.blas.get_blas_funcs('dotc', dtype=x1.dtype)
+        dot = partial(dotc, n=native(x1.size))
+    elif is_real_dtype(x1.dtype):
+        dot = np.dot  # still much faster than vdot
+    else:
+        dot = np.vdot  # slowest alternative
+    # x2 as first argument because we want linearity in x1
+    return dot(x2.data.ravel(order=x2.order),
+               x1.data.ravel(order=x1.order))
```

Comment:
Okay, that's not failure, but wrong answer. And agreed, that can happen, but not really because `x1` and `x2` are checked to be in the same space before. Anyway, they should be flattened with the same order just in case.